libzeth/circuits/mimc/mimc_mp.hpp

```diff
@@ -5,30 +5,29 @@
 #ifndef __ZETH_CIRCUITS_MIMC_MP_HPP__
 #define __ZETH_CIRCUITS_MIMC_MP_HPP__
 
-#include "libzeth/circuits/mimc/mimc.hpp"
+#include "libzeth/circuits/mimc/mimc_permutation.hpp"
 
 namespace libzeth
 {
 
-/// This gadget implements the interface of the HashTreeT template
+/// This gadget implements the interface of the HashTreeT template.
```

Comment:
Re inheriting from the permutation gadget, it wouldn't work now, but it wasn't clear to me that we would want to inherit anyway (technically the MP compression function *is* a permutation, but would we want that reflected in the types?  We might want a slightly different interface, etc.

Re inheriting from gadget, the class is still used as a gadget (in the sense that, if C++ had type-classes or traits then this would be a member of that type class), but none of the elements from the `gadget` class are used and `gadget` does not enforce any specific interface.

Reply:
> (technically the MP compression function is a permutation

How can a compression function be a permutation (i.e. the description of the re-arrangement of a set, that is bijective)?
Depending on the security assumptions employed, i.e. if the block cipher used is modelled as "Ideal" (i.e. if you rely on the "Ideal Cipher model"), then the cipher is considered as a random permutation for every key ("your key gives the permutation function"). However, if you start to build MP with the "ideal" cipher and look at the "function's domain and co-domains" you see that this is not a permutation anymore (in fact, you have a compression function)

Reply:
> How can a compression function be a permutation (i.e. the description of the re-arrangement of a set, that is bijective)?

Sorry, I phrased that badly.  Naturally, I wasn't trying to claim that "any compression function is a permutation".  I was talking about this very particular case in which the MP gadget simply calls a permutation gadget.  By looking at this code, someone might think that this MP gadget should inherit from the mimc permutation gadget (either because this particular MP might be considered a "permutation" if we reinterpret the inputs, or as an implementation strategy to reuse the permutation code, or in order that it has `libsnark::gadget` in the hierarchy), but I was arguing that it wouldn't necessarily make sense logically at the *type* level, precisely because in general the relationship would not hold.  So we would not necessarily want the caller of this *compression function* gadget to be able to use it in place of a permutation.  I'll update the comment in the code as well.

Reply:
I'v rewritten the comment to explicitly call this out.

Reply:
Thanks for rephrasing the comment @dtebbs this is clearer now! (else I'm sure I'd have raised a similar remark after looking back at the code in a few months and wondered "why do we somehow claim that a compression function is a permutation?") :)